algosdk/atomic_transaction_composer.py

```diff
@@ -165,67 +188,156 @@ def add_method_call(
+            if not approval_program or not clear_program:
+                raise error.AtomicTransactionComposerError(
+                    "One of the following required parameters for OnApplicationComplete.UpdateApplicationOC is missing: approvalProgram, clearProgram"
+                )
+            if local_schema or global_schema or extra_pages:
+                raise error.AtomicTransactionComposerError(
+                    "One of the following application creation parameters were set on a non-creation call: numGlobalInts, numGlobalByteSlices, numLocalInts, numLocalByteSlices, extraPages"
+                )
+        elif (
+            approval_program
+            or clear_program
+            or local_schema
+            or global_schema
+            or extra_pages
+        ):
+            raise error.AtomicTransactionComposerError(
+                "One of the following application creation parameters were set on a non-creation call: approvalProgram, clearProgram, numGlobalInts, numGlobalByteSlices, numLocalInts, numLocalByteSlices, extraPages"
+            )
         if not method_args:
             method_args = []
         if len(method.args) != len(method_args):
             raise error.AtomicTransactionComposerError(
                 "number of method arguments do not match the method signature"
             )
-        if not isinstance(method, abi.method.Method):
+        if not isinstance(method, abi.Method):
             raise error.AtomicTransactionComposerError(
                 "invalid Method object was passed into AtomicTransactionComposer"
             )
-        if app_id == 0:
-            raise error.AtomicTransactionComposerError(
-                "application create call not supported"
-            )
+
+        # Initialize foreign object maps
+        if not accounts:
+            accounts = []
+        if not foreign_apps:
+            foreign_apps = []
+        if not foreign_assets:
+            foreign_assets = []
 
         app_args = []
-        # For more than 14 args, including the selector, compact them into a tuple
-        additional_args = []
-        additional_types = []
+        raw_values = []
+        raw_types = []
         txn_list = []
         # First app arg must be the selector of the method
         app_args.append(method.get_selector())
         # Iterate through the method arguments and either pack a transaction
         # or encode a ABI value.
         for i, arg in enumerate(method.args):
-            if arg.type in abi.method.TRANSACTION_ARGS:
+            if abi.is_abi_transaction_type(arg.type):
                 if not isinstance(method_args[i], TransactionWithSigner):
                     raise error.AtomicTransactionComposerError(
                         "expected TransactionWithSigner as method argument, but received: {}".format(
                             method_args[i]
                         )
                     )
```

Comment:
nit: I remembered in @jasonpaulos implementation, we also need to check the `TransactionWithSigner`'s transaction match with ABI method signature's arg type, e.g.,
```js
      if (abiTypeIsTransaction(argType)) {
        if (
          !isTransactionWithSigner(argValue) ||
          !abiCheckTransactionType(argType, argValue.txn)
        ) {
```
in https://github.com/algorand/js-algorand-sdk/pull/482/files#diff-534b9601ad1de4bb83d07dc0cb5af85424ccf602bcd8f4c0a2d3b79def06e8aeR319